windows/src/desktop/kmshell/main/RemoteUpdateCheck.pas

```diff
@@ -0,0 +1,460 @@
+        Fields.Add(AnsiString('packageid_'+IntToStr(i)), AnsiString(pkg.ID));
+        Fields.Add(AnsiString('packageversion_'+IntToStr(i)), AnsiString(pkg.Version));
+        pkg := nil;
+      end;
+
+      Proxy.Server := GetProxySettings.Server;
+      Proxy.Port := GetProxySettings.Port;
+      Proxy.Username := GetProxySettings.Username;
+      Proxy.Password := GetProxySettings.Password;
+
+      Request.HostName := API_Server;
+      Request.Protocol := API_Protocol;
+      Request.UrlPath := API_Path_UpdateCheck_Windows;
+      //OnStatus :=
+      Upload;
+      if Response.StatusCode = 200 then
+      begin
+        if ucr.Parse(Response.MessageBodyAsString, 'bundle', CKeymanVersionInfo.Version) then
+        begin
+          ResponseToParams(ucr);
+
+          if FCheckOnly then
+          begin
+            // TODO: Refactor this
+            TUpdateCheckStorage.SaveUpdateCacheData(ucr);
+            Result := FParams.Result;
+          end
+          else if (Length(FParams.Packages) > 0) or (FParams.Keyman.DownloadURL <> '') then
+          begin
+            // TODO: Integrate in to the Background update state machine.
+            // for now just go straight to DownloadUpdates
+            downloadResult := DownloadUpdates;
+            if DownloadResult then
+            begin
+              Result := wucSuccess;
+            end;
+          end;
+        end
+        else
+        begin
+          FErrorMessage := ucr.ErrorMessage;
+          Result := wucFailure;
+        end;
+      end
+      else
+        raise ERemoteUpdateCheck.Create('Error '+IntToStr(Response.StatusCode));
+    finally
+      Free;
+    end;
+  except
+    on E:EHTTPUploader do
+    begin
+      if (E.ErrorCode = 12007) or (E.ErrorCode = 12029)
+        then FErrorMessage := S_OnlineUpdate_UnableToContact
+        else FErrorMessage := WideFormat(S_OnlineUpdate_UnableToContact_Error, [E.Message]);
+      Result := wucFailure;
+    end;
+    on E:Exception do
+    begin
+      FErrorMessage := E.Message;
+      Result := wucFailure;
+    end;
+  end;
+
+  with TRegistryErrorControlled.Create do  // I2890
+  try
+    if OpenKey(SRegKey_KeymanDesktop_CU, True) then
+      WriteDateTime(SRegValue_LastUpdateCheckTime, Now);
+  finally
+    Free;
+  end;
+end;
+
+function TRemoteUpdateCheck.ResponseToParams(const ucr: TUpdateCheckResponse): TRemoteUpdateCheckParams;
+var
+  i, j, n: Integer;
+  pkg: IKeymanPackage;
+begin
+  SetLength(FParams.Packages,0);
+  for i := Low(ucr.Packages) to High(ucr.Packages) do
+  begin
+    n := kmcom.Packages.IndexOf(ucr.Packages[i].ID);
+    if n >= 0 then
+    begin
+      pkg := kmcom.Packages[n];
+      j := Length(FParams.Packages);
+      SetLength(FParams.Packages, j+1);
+      FParams.Packages[j].NewID := ucr.Packages[i].NewID;
+      FParams.Packages[j].ID := ucr.Packages[i].ID;
+      FParams.Packages[j].Description := ucr.Packages[i].Name;
+      FParams.Packages[j].OldVersion := pkg.Version;
```

Comment:
This looks like it is setting this to the installed version vs the "previous" version that would be in ucr.Packages.[i].OldVersion.   We could update FParams to be of type `TUpdateCheckResponse` and set the the OldVersion to the Package version but applying to different meanings dependent on use may provide future confusion.  We could just look up the pkg.Version when we go to use it `UpdateXMLRender`